libmamba/include/mamba/util/flat_bool_expr_tree.hpp

```diff
@@ -0,0 +1,738 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_UTIL_FLAT_EXPR_TREE_HPP
+#define MAMBA_UTIL_FLAT_EXPR_TREE_HPP
+
+#include <functional>
+#include <utility>
+#include <variant>
+#include <vector>
+
+namespace mamba::util
+{
+    /**
+     * In array binary tree.
+     *
+     * A binary tree, each node is either a leaf, or a node with exactly two children.
```

Comment:
It's not clear to me if this is supposed to be a container like `flat_set` in boost (or the standard one but it's 2 vectors) or more like a logical state thing?

Reply:
We already have a flat_set.
This is a binary tree meant to be used as a binary tree, *i.e.* not a tree in the sense *the implementation of a set-like interface based on a tree*. The goal is to query who is a child of whom, not make element unique (they don't have to).

It is flat in the sense that the nodes are all allocated in a vector instead of individually.

Reply:
Ok that looks useful indeed!